File: dotnet/JourneyBook.Infrastructure/Tiles/TileCache.cs
namespace JourneyBook.Infrastructure.Tiles;

/// <summary>
/// Disk-backed tile cache under a single root (default <c>data/cache</c>), keyed
/// <c>{source}/{z}/{x}/{y}.{ext}</c>. The store records the real extension (png/pbf/…); the
/// lookup discovers whichever <c>{y}.*</c> exists, so a tile cached as <c>pbf</c> is found on the
/// next request without the caller having to know the type up front. Every path is confined to
/// the root using the same resolved-path <c>StartsWith</c> idiom as
/// <c>GeneratedPdfService.TryDeleteArtifact</c>: a key that escapes the root (e.g. <c>../</c>) is
/// treated as a miss / skipped store, never written. Stores are atomic (temp file + rename) so a
/// concurrent reader never sees a torn tile.
/// </summary>
public sealed class TileCache
{
    private readonly string _root;

    public TileCache(string cacheRoot)
    {
        _root = Path.GetFullPath(string.IsNullOrWhiteSpace(cacheRoot) ? "data/cache" : cacheRoot);
    }

    /// <summary>True on a hit; <paramref name="ext"/> is the discovered file extension (no dot).</summary>
    public bool TryGet(string sourceKey, int z, int x, int y, out byte[] bytes, out string ext)
    {
        bytes = [];
        ext = string.Empty;

        if (!TryResolveDir(sourceKey, z, x, out var dir) || !Directory.Exists(dir))
        {
            return false;
        }

        try
        {
            var match = Directory.EnumerateFiles(dir, $"{y}.*").FirstOrDefault();
            if (match is null)
            {
                return false;
            }

            bytes = File.ReadAllBytes(match);
            ext = Path.GetExtension(match).TrimStart('.');
            return true;
        }
        catch
        {
            bytes = [];
            ext = string.Empty;
            return false;
        }
    }

    public void Store(string sourceKey, int z, int x, int y, string ext, byte[] bytes)
    {
        if (!TryResolveDir(sourceKey, z, x, out var dir))
        {
            // Escaping key: skip rather than write outside the cache root.
            return;
        }

        try
        {
            Directory.CreateDirectory(dir);
            var path = Path.Combine(dir, $"{y}.{ext}");
            var temp = path + ".tmp-" + Guid.NewGuid().ToString("N");
            File.WriteAllBytes(temp, bytes);
            File.Move(temp, path, overwrite: true);
        }
        catch
        {
            // Best-effort: a cache write failure must not break tile serving.
        }
    }

    /// <summary>
    /// Builds the <c>{root}/{source}/{z}/{x}</c> directory and confirms it resolves within the
    /// cache root. Returns false (and an empty path) when the composed path would escape the root.
    /// </summary>
    private bool TryResolveDir(string sourceKey, int z, int x, out string dir)
    {
        dir = string.Empty;
        var candidate = Path.GetFullPath(Path.Combine(_root, sourceKey, z.ToString(), x.ToString()));
        var rootWithSep = _root.EndsWith(Path.DirectorySeparatorChar) ? _root : _root + Path.DirectorySeparatorChar;
        if (!candidate.StartsWith(rootWithSep, StringComparison.Ordinal))
        {
            return false;
        }

        dir = candidate;
        return true;
    }
}

File: packages/map-sources/src/tilecache.ts
/**
 * Minimal Node-side disk tile cache, honoring the same `{source}/{z}/{x}/{y}.{ext}` key as the
 * C# proxy cache (so the headless CLI and the API share one cache directory). The store records
 * the real extension; the lookup discovers whichever `{y}.*` exists. Paths are confined to the
 * cache root (a `../` key resolves outside and is treated as a miss / skipped store). No TTL —
 * eviction is a Stage 7 concern.
 */

import { promises as fs } from "node:fs";
import * as path from "node:path";

/**
 * Cache file extension for a tile's `Content-Type`.
 *
 * The C# proxy (`TileService.ExtFor`) and this module write into the SAME
 * directory layout — `{source}/{z}/{x}/{y}.{ext}` — and each reads back what the
 * other stored, so the extension is a shared contract, not a local choice. The
 * store site in `panel.ts` used to pass a literal `"png"` for every tile it had
 * just fetched, so a JPEG cached by the headless CLI was filed as `.png` and the
 * proxy then served those bytes as `image/png` (`ContentTypeFor("png")`).
 *
 * `data/fixtures/tile-content-types.json` is the table both languages are tested
 * against; this function must not be edited without it.
 *
 * A content type with parameters (`image/jpeg; charset=binary`) is normalised to
 * its media type first — the C# switch matches the bare string and would fall to
 * its `png` default there. That is a superset, not a divergence: every bare type
 * maps identically, and the parameterised form is one the C# path gets wrong.
 */
export function tileExtensionForContentType(contentType: string | null | undefined): string {
  const media = (contentType ?? "").split(";")[0]!.trim().toLowerCase();
  switch (media) {
    case "application/x-protobuf":
    case "application/vnd.mapbox-vector-tile":
      return "pbf";
    case "image/jpeg":
      return "jpg";
    case "image/webp":
      return "webp";
    default:
      // Unknown or absent: png, matching the C# default arm. Guessing wrong is
      // recoverable (the bytes are intact and re-fetchable); refusing to cache
      // a tile because its source omitted a header is not.
      return "png";
  }
}

/** Inverse of {@link tileExtensionForContentType}; mirrors `TileService.ContentTypeFor`. */
export function contentTypeForTileExtension(ext: string): string {
  switch (ext.toLowerCase()) {
    case "pbf":
      return "application/x-protobuf";
    case "jpg":
    case "jpeg":
      return "image/jpeg";
    case "webp":
      return "image/webp";
    default:
      return "image/png";
  }
}

function resolveTileDir(cacheDir: string, source: string, z: number, x: number): string | null {
  const root = path.resolve(cacheDir);
  const dir = path.resolve(root, source, String(z), String(x));
  const rootWithSep = root.endsWith(path.sep) ? root : root + path.sep;
  if (!dir.startsWith(rootWithSep)) {
    return null; // escapes the cache root
  }
  return dir;
}

/**
 * True for a finished cache entry named `{y}` or `{y}.{ext}` and nothing else.
 *
 * A `startsWith("{y}.")` prefix test also matches the `{y}.{ext}.tmp-{pid}-{ms}`
 * file storeCachedTile writes before its rename, so a concurrent reader would be
 * served a half-written tile as a hit — defeating the atomicity the temp-file
 * dance exists for. A single alphanumeric extension segment is the whole rule.
 */
function isTileName(name: string, y: number): boolean {
  return name === `${y}` || new RegExp(`^${y}\\.[A-Za-z0-9]+$`).test(name);
}

/** Returns the cached tile bytes + discovered extension, or null on a miss. */
export async function getCachedTile(
  cacheDir: string,
  source: string,
  z: number,
  x: number,
  y: number,
): Promise<{ bytes: Buffer; ext: string } | null> {
  const dir = resolveTileDir(cacheDir, source, z, x);
  if (dir === null) return null;

  try {
    const entries = await fs.readdir(dir);
    const match = entries.find((f) => isTileName(f, y));
    if (!match) return null;
    const bytes = await fs.readFile(path.join(dir, match));
    const ext = path.extname(match).replace(/^\./, "");
    return { bytes, ext };
  } catch {
    return null;
  }
}

/** Writes a tile to the cache (atomic temp-file + rename). No-op if the key escapes the root. */
export async function storeCachedTile(
  cacheDir: string,
  source: string,
  z: number,
  x: number,
  y: number,
  ext: string,
  bytes: Buffer,
): Promise<void> {
  const dir = resolveTileDir(cacheDir, source, z, x);
  if (dir === null) return;

  try {
    await fs.mkdir(dir, { recursive: true });
    const final = path.join(dir, `${y}.${ext}`);
    const temp = `${final}.tmp-${process.pid}-${Date.now()}`;
    await fs.writeFile(temp, bytes);
    await fs.rename(temp, final);
  } catch {
    // best-effort: a cache write failure must not break rendering
  }
}
